### PR TITLE
BSON Handlers improvements

### PR DESCRIPTION
**Pull request by jtjeferreira**
# Pull Request Checklist

* [X] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [X] Have you read through the [contributor guidelines](https://github.com/ReactiveMongo/ReactiveMongo/blob/master/CONTRIBUTING.md#reactivemongo-developer--contributor-guidelines)?
* [X] Have you added tests for any changed functionality? No changed functionality and current test still pass

## Purpose

improves compilation speed by making implicit search for BSONReaders and BSONWriters faster

## Background Context

This is the same approach used in other libraries (play-json, circe)

### CODE REVIEW

File: bson/src/main/scala/DefaultBSONHandlers.scala [hunk 1]
```diff
@@ -120,24 +120,6 @@ trait DefaultBSONHandlers {
   implicit def bsonBooleanLikeReader[B <: BSONValue] =
     new BSONBooleanLikeReader[B]
 
-  // Collections Handlers
```
Comment (cchantep) [inline bson/src/main/scala/DefaultBSONHandlers.scala]: I don't see the goal, those instances are not low priority for me

Comment (jtjeferreira) [reply bson/src/main/scala/DefaultBSONHandlers.scala]: I can prepare some examples with "-Xlog-implicits" so you can see the compiler trying to use BSONArrayCollectionWriter where it is not needed, but like I said above other libraries to it like this: [play-json](https://github.com/playframework/play-json/blob/master/play-json/shared/src/main/scala/play/api/libs/json/Reads.scala#L186) and [circe](https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L281)

Comment (cchantep) [reply bson/src/main/scala/DefaultBSONHandlers.scala]: There are other reason why Play JSON is doing it in this way (I'm contrib on both).

**Comment by jtjeferreira** [comment]
I will prepare the examples then. Maybe later today...

File: bson/src/main/scala/DefaultBSONHandlers.scala (hunk 1 above)
Comment (jtjeferreira) [reply bson/src/main/scala/DefaultBSONHandlers.scala]: I will prepare the examples then. Maybe later today

Comment (cchantep) [reply bson/src/main/scala/DefaultBSONHandlers.scala]: thx

File: bson/src/main/scala/DefaultBSONHandlers.scala [hunk 2]
```diff
@@ -184,9 +166,6 @@ trait DefaultBSONHandlers {
   implicit def findWriter[T](implicit writer: VariantBSONWriter[T, _ <: BSONValue]): BSONWriter[T, _ <: BSONValue] =
     new VariantBSONWriterWrapper(writer)
 
-  implicit def findReader[T](implicit reader: VariantBSONReader[_ <: BSONValue, T]): BSONReader[_ <: BSONValue, T] =
```
Comment (jtjeferreira) [inline bson/src/main/scala/DefaultBSONHandlers.scala]: please note that I also remove this implicit, because I think it is never used...

**Pull request merged**